password.c

```diff
@@ -0,0 +1,227 @@
+
+/* The password setting */
+const char* PASSWORD_SETTING = "PASSWD_CMDS";
+
+/* Regex list */
+REGEX_NODE *global_regex_list = NULL;
+
+/* Append regex to list */
+int append_regex_to_list(regex_t regex)
+{
+    /* Create and initialize regex node */
+    REGEX_NODE *new_regex_node = (REGEX_NODE *)malloc(sizeof(REGEX_NODE));
+    if (new_regex_node == NULL)
+    {
+        /* When allocate memory failed, stop and return. also output log to both syslog and stderr with LOG_PERROR*/
+        trace("Failed to allocate memory for regex node.\n");
+        return REGEX_APPEND_FAILED;
+    }
+
+    new_regex_node->next = NULL;
+    new_regex_node->regex = regex;
+
+    /* Find the pointer to the latest regex node's 'next' field */
+    REGEX_NODE **current_node = &global_regex_list;
+    while (*current_node != NULL) {
+        current_node = &((*current_node)->next);
+    }
+    
+    /* append new regex to tail node */
+    *current_node = new_regex_node;
+    return REGEX_APPEND_SUCCESS;
+}
+
+/* Release password setting */
+void release_password_setting()
+{
+    if (global_regex_list == NULL) {
+        return;
+    }
+
+    /* Walk to last regex */
+    REGEX_NODE *current = global_regex_list;
+    while (current != NULL) {
+        /* Continue with next regex */
+        REGEX_NODE* current_node_memory = current;
+        current = current->next;
+        
+        /* Free node memory, this may also reset all allocated memory depends on c lib implementation */
+        free(current_node_memory);
+    }
+
+    /* Reset list */
+    global_regex_list = NULL;
+}
+
+/* Replace password with regex */
+void remove_password(char* command)
+{
+    if (global_regex_list == NULL) {
+        return;
+    }
+
+    /* Check every regex */
+    REGEX_NODE *next_node = global_regex_list;
+    while (next_node != NULL) {
+        /* Try fix password with current regex */
+        if (remove_password_by_regex(command, next_node->regex) == PASSWORD_REMOVED) {
+            return;
+        }
+        
+        /* If password not fix, continue try next regex */
+        next_node = next_node->next;
+    }
+}
+
+/* Find and return the pointer of the first none space character*/
+char* find_non_space(char *str)
+{
+    if (str == NULL) {
+        return str;
+    }
+
+    while (isspace(*str)) {
+        str++;
+    }
+
+    return str;
+}
+
+/* Check setting if have next line */
+int check_have_next_line(const char *str)
+{
+    if (str == NULL) {
+        return NO_NEXT_SETTING_LINE;
+    }
+
+    /* Find end of string */
+    const char* endpos = str;
+    while (*endpos) {
+        endpos++;
+    }
+
+    /* Find last none whitespace character */
+    char last_none_whitespace_char = 0;
+    while (endpos-- > str) {
+        if (!isspace(*endpos)) {
+            last_none_whitespace_char = *endpos;
+            break;
+        }
+    }
+
+    /* If the string end with \, then have next setting line */
+    if (last_none_whitespace_char == '\\') {

```

Comment:
>\\ [](http://example.com/codeflow?start=38&length=2)

The doc says
> Long lines can be continued with a backslash (‘\’) as the last character on the line

However you are accepting single backslash following spaces.

Reply:
Fixed.
Create sudoers_helper.c and following method will load PASSWD_CMDS from sudoers file:

char* load_passwd_cmds(const char *setting_path)